### PR TITLE
Optional constraints for Input<Func> and Output<Func>

### PR DESCRIPTION
**Pull request by steven-johnson**
This is a sketch of a possible way to address Issue #3306: `Input<Func>` and `Output<Func>` have a new method, `dimensions_and_alignment()`, which returns a struct that lets you get/set the min/extent/stride/host-alignment; these values are entirely ignored if the given Input/Output isn't AOT-compiled as a Buffer.

This works fine and I feel it's robust, but I'm not sure if this API is one I'm happy to support forever. That said, it would make it a lot easier to make composable Generators per the details in the issue.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Attn @inazarenko

**Comment by abadams** [comment]
Can we not just support output.dim(2).set_stride(...) and co. on Output<Func> ?

**Comment by steven-johnson** [comment]
> Can we not just support output.dim(2).set_stride(...) and co. on Output<Func> ?

No, because the `Output<Func>` might not have any realization if the Generator has been inlined into another Generator.

**Comment by abadams** [comment]
Why can't output.dim(1).set_stride mean the same thing as output.dimensions_and_alignment().dim(1).set_stride? I.e. it gets ignored if there's no realization.

**Comment by steven-johnson** [comment]
> gets ignored if there's no realization

The thing I'm concerned about is that it's easy to use `dim(0).extent()`, etc mistakenly, and then get weird or incorrect results that are hard to debug. Let me see if I can reliably ensure that using these inappropriately will produce a hard+clear failure at compile time.

**Comment by dsharletg** [comment]
Maybe a bit pie in the sky, but I'd love to try to reduce/eliminate the inconsistency between internal and output stages in general, which might implicitly make this problem better.

For example, it seems like we should be able to use `bound` and `reorder_storage` to implement interleaved output buffers, the way you would for an internal stage. The fact that we have a different language for describing this for internal and external stages is unfortunate IMO and a source of confusion for myself at least.

Basically, I think it would be awesome to get rid of set_min/extent_stride completely, and use reorder_storage, align_bounds, and bound on output stages instead.

**Comment by abadams** [comment]
+1

**Comment by inazarenko** [comment]
> That said: will all of these work on Funcs that may be inlined? 

In particular, Funcs wrapped around buffers in the implementation of `Input<Func>` are always inlined.
